shell/src/main/java/alluxio/shell/command/CpCommand.java

```diff
@@ -0,0 +1,236 @@
+
+  /**
+   * @param conf the configuration for Alluxio
+   * @param fs the filesystem of Alluxio
+   */
+  public CpCommand(Configuration conf, FileSystem fs) {
+    super(conf, fs);
+  }
+
+  @Override
+  public String getCommandName() {
+    return "cp";
+  }
+
+  @Override
+  protected int getNumOfArgs() {
+    return 2;
+  }
+
+  @Override
+  public void run(CommandLine cl) throws IOException {
+    String[] args = cl.getArgs();
+    AlluxioURI srcPath = new AlluxioURI(args[0]);
+    AlluxioURI dstPath = new AlluxioURI(args[1]);
+    List<AlluxioURI> srcPaths = AlluxioShellUtils.getAlluxioURIs(mFileSystem, srcPath);
+    if (srcPaths.size() == 0) {
+      throw new IOException(srcPath.getPath() + " does not exist.");
+    }
+
+    if (srcPath.containsWildcard()) {
+      copyWildcard(srcPaths, dstPath);
+    } else {
+      copy(srcPath, dstPath);
+    }
+  }
+
+  /**
+   * Copies a list of files or directories specified by srcPaths to the destination specified by
+   * dstPath. This method is used when the original source path contains wildcards.
+   *
+   * @param srcPaths a list of files or directories in the Alluxio filesystem
+   * @param dstPath the destination in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copyWildcard(List<AlluxioURI> srcPaths, AlluxioURI dstPath) throws IOException {
+    URIStatus dstStatus = null;
+    try {
+      dstStatus = mFileSystem.getStatus(dstPath);
+    } catch (FileDoesNotExistException e) {
+      // if the destination does not exist, it will be created
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    if (dstStatus != null && !dstStatus.isFolder()) {
+      throw new IOException(
+          ExceptionMessage.DESTINATION_FILE_CANNOT_EXIST_WITH_WILDCARD_SOURCE.getMessage());
+    }
+    if (dstStatus == null) {
+      try {
+        mFileSystem.createDirectory(dstPath);
+        System.out.println("Created directory: " + dstPath.getPath());
+      } catch (AlluxioException e) {
+        throw new IOException("Fail to create directory: " + dstPath.getPath());
+      }
+    }
+    List<String> errorMessages = new ArrayList<>();
+    for (AlluxioURI srcPath : srcPaths) {
+      try {
+        copy(srcPath, new AlluxioURI(dstPath.getScheme(), dstPath.getAuthority(),
+            PathUtils.concatPath(dstPath.getPath(), srcPath.getName())));
+      } catch (IOException e) {
+        errorMessages.add(e.getMessage());
+      }
+    }
+    if (errorMessages.size() != 0) {
+      throw new IOException(Joiner.on('\n').join(errorMessages));
+    }
+  }
+
+  /**
+   * Copies a file or a directory in the Alluxio filesystem.
+   *
+   * @param srcPath the source {@link AlluxioURI} (could be a file or a directory)
+   * @param dstPath the destination path in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copy(AlluxioURI srcPath, AlluxioURI dstPath) throws IOException {
```

Comment:
after addressing an earlier comment the duplicate code is small enough that I don't think it is worth the boilerplate to factor this out